ruma-events/src/pdu.rs

```diff
@@ -0,0 +1,934 @@
+            auth_events: self.auth_events,
+            redacts: self.redacts,
+            unsigned: self.unsigned,
+            hashes: self.hashes,
+            signatures: self.signatures,
+        }
+    }
+}
+
+/// Stub for PDUs of room versions 3 and above.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RoomV3PduStub {
+    /// The user id of the user who sent this event.
+    pub sender: UserId,
+
+    /// The `server_name` of the homeserver that created this event.
+    pub origin: String,
+
+    /// Timestamp (milliseconds since the UNIX epoch) on originating homeserver
+    /// of when this event was created.
+    #[serde(with = "ruma_serde::time::ms_since_unix_epoch")]
+    pub origin_server_ts: SystemTime,
+
+    // TODO: Replace with event content collection from ruma-events once that exists
+    /// The event's type.
+    #[serde(rename = "type")]
+    pub kind: EventType,
+
+    /// The event's content.
+    pub content: JsonValue,
+
+    /// A key that determines which piece of room state the event represents.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub state_key: Option<String>,
+
+    /// Event IDs for the most recent events in the room that the homeserver was
+    /// aware of when it created this event.
+    pub prev_events: Vec<EventId>,
+
+    /// The maximum depth of the `prev_events`, plus one.
+    pub depth: UInt,
+
+    /// Event IDs for the authorization events that would allow this event to be
+    /// in the room.
+    pub auth_events: Vec<EventId>,
+
+    /// For redaction events, the ID of the event being redacted.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub redacts: Option<EventId>,
+
+    /// Additional data added by the origin server but not covered by the
+    /// signatures.
+    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
+    pub unsigned: BTreeMap<String, JsonValue>,
+
+    /// Content hashes of the PDU.
+    pub hashes: EventHash,
+
+    /// Signatures for the PDU.
+    pub signatures: BTreeMap<String, BTreeMap<String, String>>,
+}
+
+impl RoomV3PduStub {
+    /// Converts a V3 PDU stub into a full V3 PDU.
+    pub fn into_v3_pdu(self, room_id: RoomId) -> RoomV3Pdu {
+        RoomV3Pdu {
+            room_id,
+            sender: self.sender,
+            origin: self.origin,
+            origin_server_ts: self.origin_server_ts,
+            kind: self.kind,
+            content: self.content,
+            state_key: self.state_key,
+            prev_events: self.prev_events,
+            depth: self.depth,
+            auth_events: self.auth_events,
+            redacts: self.redacts,
+            unsigned: self.unsigned,
+            hashes: self.hashes,
+            signatures: self.signatures,
+        }
+    }
+}
+
+/// Content hashes of a PDU.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct EventHash {
+    /// The SHA-256 hash.
+    pub sha256: String,
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
With the tests being 2/3 the size of the module, could you move them outside the source into a separate file in `ruma-events/tests`?